saleor/payment/migrations/tasks/saleor3_12.py

```diff
@@ -0,0 +1,312 @@
+from decimal import Decimal
+
+from django.db.models import F, Q, Case, OuterRef, Subquery, Exists, Func, When
+
+from ....celeryconf import app
+from ...models import TransactionEvent, TransactionItem
+
+# Batch size of 1000 is about ~20MB of memory usage in task
+BATCH_SIZE = 1000
+
+
+def _get_ids(qs, batch_size=None):
+    if batch_size:
+        return qs.values_list("pk", flat=True)[:batch_size]
+    return qs.values_list("pk", flat=True)
+
+
+def _create_transaction_events(transaction_event_class, transactions_qs, type):
+    transaction_events = []
+    for pk, amount, amount_sum, currency in transactions_qs:
+        amount_sum = 0 if not amount_sum else amount_sum
+        transaction_events.append(
+            transaction_event_class(
+                transaction_id=pk,
+                type=type,
+                amount_value=amount - amount_sum,
+                currency=currency,
+                include_in_calculations=True,
+                message="Manual adjustment of the transaction.",
+            )
+        )
+    if transaction_events:
+        transaction_event_class.objects.bulk_create(transaction_events)
+
+
+def _get_events(transaction_event_class, type):
+    return (
+        transaction_event_class.objects.filter(
+            transaction=OuterRef("pk"), type=type, include_in_calculations=True
+        )
+        .annotate(amount_sum=Func(F("amount_value"), function="Sum"))
+        .values("amount_sum")
+    )
+
+
+def create_event_for_authorized(
+    transaction_item_class, transaction_event_class, batch_size=None
+):
+    transaction_events = []
+    authorize_events = transaction_event_class.objects.filter(
+        Q(transaction=OuterRef("pk"))
+        & (Q(type="authorization_success") | Q(type="authorization_adjustment"))
+        & Q(include_in_calculations=True)
+    )
```

Comment:
Shouldn't we have ordering by date in all querysets here so that we always start processing from the newest items?